### PR TITLE
SCP-2285: Remove ExMemory annotations from CekValue

### PR DESCRIPTION
**Pull request by michaelpj**
Easy enough, and it nets us a modest win.

```
crowdfunding/1            1.467 ms       1.443 ms       -1.6%
crowdfunding/2            1.462 ms       1.441 ms       -1.4%
crowdfunding/3            1.457 ms       1.439 ms       -1.2%
crowdfunding/4            706.4 μs       691.4 μs       -2.1%
crowdfunding/5            709.5 μs       692.5 μs       -2.4%
future/1                  833.7 μs       812.0 μs       -2.6%
future/2                  1.729 ms       1.693 ms       -2.1%
future/3                  1.727 ms       1.690 ms       -2.1%
future/4                  2.158 ms       2.079 ms       -3.7%
future/5                  2.642 ms       2.578 ms       -2.4%
future/6                  2.201 ms       2.142 ms       -2.7%
future/7                  2.638 ms       2.586 ms       -2.0%
multisigSM/1              1.540 ms       1.512 ms       -1.8%
multisigSM/2              1.561 ms       1.541 ms       -1.3%
multisigSM/3              1.583 ms       1.562 ms       -1.3%
multisigSM/4              1.601 ms       1.589 ms       -0.7%
multisigSM/5              1.903 ms       1.888 ms       -0.8%
multisigSM/6              1.533 ms       1.510 ms       -1.5%
multisigSM/7              1.558 ms       1.541 ms       -1.1%
multisigSM/8              1.591 ms       1.562 ms       -1.8%
multisigSM/9              1.613 ms       1.579 ms       -2.1%
multisigSM/10             1.916 ms       1.880 ms       -1.9%
vesting/1                 1.490 ms       1.442 ms       -3.2%
vesting/2                 1.298 ms       1.269 ms       -2.2%
vesting/3                 1.489 ms       1.443 ms       -3.1%
marlowe/trustfund/1       4.134 ms       4.030 ms       -2.5%
marlowe/trustfund/2       3.031 ms       3.000 ms       -1.0%
marlowe/zerocoupon/1      4.224 ms       4.176 ms       -1.1%
marlowe/zerocoupon/2      2.673 ms       2.630 ms       -1.6%
```

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs [hunk 1]
```diff
@@ -462,7 +462,7 @@ lookupVarName varName varEnv = do
 -- | The entering point to the CEK machine's engine.
 enterComputeCek
     :: forall uni fun s
-    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s, uni `Everywhere` ExMemoryUsage)
     => CekMachineCosts
     -> Context uni fun
     -> CekValEnv uni fun
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:468]: So this still receives a `TermWithMem`? Can we drop `memTerm = withMemory term` from `runCek` and then erase all occurrences of `TermWithMem` completely?

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:468]: I was going to say "I'll do that separately"... but actually I think it is indeed totally dead now, so I might as well remove those annotations too.

**Comment by michaelpj** [comment]
I murdered the ExMemory annotations on terms too. I'll re-benchmark once Kenneth is done with the machine, I want to know how much difference that makes on top.

**Review by kwxm — APPROVED** [review]
Yup. Go for it.

**Comment by michaelpj** [comment]
Kenneth beat me to it, it's great:
```
crowdfunding/1            1.444 ms	 1.062 ms       -26.5%
crowdfunding/2            1.437 ms	 1.062 ms       -26.1%
crowdfunding/3            1.435 ms	 1.062 ms       -26.0%
crowdfunding/4            689.5 μs	 463.0 μs       -32.8%
crowdfunding/5            689.0 μs	 465.8 μs       -32.4%
future/1                  809.6 μs	 571.8 μs       -29.4%
future/2                  1.690 ms	 1.296 ms       -23.3%
future/3                  1.697 ms	 1.302 ms       -23.3%
future/4                  2.092 ms	 1.609 ms       -23.1%
future/5                  2.596 ms	 1.985 ms       -23.5%
future/6                  2.155 ms	 1.575 ms       -26.9%
future/7                  2.598 ms	 1.981 ms       -23.7%
multisigSM/1              1.507 ms	 1.052 ms       -30.2%
multisigSM/2              1.538 ms	 1.052 ms       -31.6%
multisigSM/3              1.559 ms	 1.073 ms       -31.2%
multisigSM/4              1.578 ms	 1.091 ms       -30.9%
multisigSM/5              1.879 ms	 1.340 ms       -28.7%
multisigSM/6              1.503 ms	 1.050 ms       -30.1%
multisigSM/7              1.532 ms	 1.052 ms       -31.3%
multisigSM/8              1.555 ms	 1.076 ms       -30.8%
multisigSM/9              1.573 ms	 1.098 ms       -30.2%
multisigSM/10             1.887 ms	 1.348 ms       -28.6%
vesting/1                 1.448 ms	 1.035 ms       -28.5%
vesting/2                 1.273 ms	 883.7 μs       -30.6%
vesting/3                 1.448 ms	 1.035 ms       -28.5%
marlowe/trustfund/1       4.011 ms	 3.091 ms       -22.9%
marlowe/trustfund/2       2.985 ms	 2.184 ms       -26.8%
marlowe/zerocoupon/1      4.148 ms	 3.394 ms       -18.2%
marlowe/zerocoupon/2      2.613 ms	 1.857 ms       -28.9%
```

**Comment by michaelpj** [comment]
Which is still super weird, I don't understand how `memTerm` can be so expensive.

**Comment by michaelpj** [comment]
Maybe it really was carrying around all those extra integers in the AST? Who knows.

**Comment by kwxm** [comment]
> Which is still super weird, I don't understand how `memTerm` can be so expensive.

Also there's the (lazy) traversal of the AST to add the annotations at the start.  That's the thing that was massively expensive when it used `<$`.

**Comment by kwxm** [comment]
> Maybe it really was carrying around all those extra integers in the AST? Who knows.

At some point I'll try this branch on the nofib benchmarks.  They take a lot longer to run, so that might reveal if we're dealing with a constant(ish) overhead or if there's some omnipresent slowdown.

**Comment by kwxm** [comment]
This is interesting.  I ran the nofib benchmarks on the last commit of this branch at `a3e6ac6` with master at `3ec697966` and got a slowdown!

```
clausify/formula1         66.57 ms	 72.88 ms       +9.5%
clausify/formula2         82.63 ms	 90.79 ms       +9.9%
clausify/formula3         230.1 ms	 252.0 ms       +9.5%
clausify/formula4         318.9 ms	 348.1 ms       +9.2%
clausify/formula5         1.485 s 	 1.636 s        +10.2%
knights/4x4               181.0 ms	 192.2 ms       +6.2%
knights/6x6               514.3 ms	 556.2 ms       +8.1%
knights/8x8               857.9 ms	 934.1 ms       +8.9%
primetest/05digits        125.0 ms	 133.5 ms       +6.8%
primetest/08digits        230.0 ms	 245.1 ms       +6.6%
primetest/10digits        324.0 ms	 344.4 ms       +6.3%
primetest/20digits        646.8 ms	 688.0 ms       +6.4%
primetest/30digits        936.9 ms	 996.1 ms       +6.3%
primetest/40digits        1.254 s 	 1.335 s        +6.5%
primetest/50digits        1.239 s 	 1.319 s        +6.5%
queens4x4/bt              30.44 ms	 31.98 ms       +5.1%
queens4x4/bm              39.64 ms	 41.61 ms       +5.0%
queens4x4/bjbt1           37.59 ms	 39.36 ms       +4.7%
queens4x4/bjbt2           40.09 ms	 42.23 ms       +5.3%
queens4x4/fc              92.41 ms	 98.73 ms       +6.8%
queens5x5/bt              410.7 ms	 437.2 ms       +6.5%
queens5x5/bm              450.1 ms	 476.5 ms       +5.9%
queens5x5/bjbt1           482.0 ms	 510.4 ms       +5.9%
queens5x5/bjbt2           508.5 ms	 540.1 ms       +6.2%
queens5x5/fc              1.176 s 	 1.260 s        +7.1%
```

Just to be sure I reran the validation benchmarks on those commits and got > 25% reductions in times, as in Michael's table above.  I'll have another go once all of today's PRs are merged.

**Comment by effectfully** [comment]
> This is interesting. I ran the nofib benchmarks on the last commit of this branch at a3e6ac6 with master at 3ec697966 and got a slowdown!

So in the end we were caching something indeed? Anyway, clearly we should prioritize validation benchmarks and the speed-up is huge there.

**Comment by kwxm** [comment]
> Anyway, clearly we should prioritize validation benchmarks and the speed-up is huge there.

Absolutely!  It's all a bit mysterious though. Maybe the initial traversal was indeed expensive but for longer-running programs the gains from caching compensated for that (although are we _really_ looking up that many literal constants)?  Anyway, I agree that we should take what we can get for validation.

**Pull request merged**